### PR TITLE
Prod image

### PR DESCRIPTION
**Pull request by princemaple**
- [x] get it to build
- [x] get it to run
- [x] sync cli supports somehow
- [x] explore minimum size - 74MB on alpine

Refs #260 

---

> Say we have `firezone/firezone` on Docker Hub.

## How to deploy

Copy `docker-compose.prod.yml` to your server, or your machine / jump box. Rename it to `docker-compose.yml` if you don't want to type `-f docker-compose.prod.yml` all the time.

### On-Server Deployment

1. `docker run --rm firezone/firezone bin/gen-env > .env`,
    - doesn't have to be `.env`, will talk about about this below
1. modify the generated file with correct admin email, external url and wireguard endpoint
    - optionally change the rest of passwords / secrets if preferred
1. `docker compose -f docker-compose.prod.yml up -d`
    - if you used a different name from `.env`, you can pass `--env-file YOUR_ENV_FILE` *before* `up`, i.e. `docker compose -f docker-compose.prod.yml --env-file MY_ENV up -d`
1. `docker compose -f docker-compose.prod.yml exec firezone bin/create-or-reset-admin`

Now you have a running firezone app with auto https enabled

### Swarm deployment (on your machine or jump box)

> Given you have a docker context `my-server` that points to the deployment destination

1. (same) `docker run --rm firezone/firezone bin/gen-env > .env`,
    - doesn't have to be `.env`, will talk about about this below
1. (same) modify the generated file with correct admin email, external url and wireguard endpoint
    - optionally change the rest of passwords / secrets if preferred
1. `env $(cat .env) docker --context my-server stack deploy -c docker-compose.prod.yml my-firezone`

### Custom

Pull down `firezone/firezone` image and go nuts. Compose your own services or stuck it in your k8s cluster. It's up to you.

## Commands

ctl commands are not really needed anymore other than `create-or-reset-admin`. Network changes are contained inside the containers. HTTPS is fully automated by caddy, it takes care of certificate revocation before expiry date as well. And kind of reset / update is just a re-deploy.

## Future works

Right now it's like MVP for containerization. Future works involve reducing reliance on files (e.g. private key file) and provide option to load from env. Provide more documentation around High Availability. Provide docs on other orchestrators, namely k8s and maybe podman.

### CODE REVIEW

**Comment by jamilbk** [comment]
Looks like [there's a hexpm alpine image](https://hub.docker.com/layers/elixir/hexpm/elixir/1.13.4-erlang-25.0.3-alpine-3.16.0/images/sha256-1a662d002f12c9f1d24cf284a52a5460e8bcc2bc11d1d0d2ac13f7b01135098a?context=explore) if that will help reduce size at all.

**Comment by princemaple** [comment]
I will try the Alpine version after finishing up all the functionality. I'm not sure how well alpine will work with wireguard and nft. It would reduce at most 25MB if it worked.

**Comment by jamilbk** [comment]
> I will try the Alpine version after finishing up all the functionality. I'm not sure how well alpine will work with wireguard and nft. It would reduce at most 25MB if it worked.

K cool. I don't _think_ there should be any runtime dependencies on wireguard anymore since Wireguardex should just interface directly with the kernel module on the host. `nft` _should_ be available with `apk add nftables` but curious to see what you find.

If you're planning to publish the image in CI like the Omnibus package is published, you can use the following Github secrets with the [docker build action](https://github.com/marketplace/actions/build-and-push-docker-images) which will be populated for this repo:

```
username: ${{ secrets.DOCKERHUB_USERNAME }}
password: ${{ secrets.DOCKERHUB_TOKEN }}
```

**Comment by princemaple** [comment]
Thanks! Nice preparation.

**Comment by princemaple** [comment]
Update for the week: I've managed to run prod image with a compose file on a server.

**Comment by jamilbk** [comment]
🔥

File: docker-compose.prod.yml
Comment (princemaple) [inline docker-compose.prod.yml:30]:
```suggestion
    image: firezone/firezone
```

Waiting on confirmation of docker hub repo.

File: Dockerfile.prod
Comment (jamilbk) [inline Dockerfile.prod]: Is there a way to enable telemetry by default?

File: docker-compose.prod.yml
Comment (jamilbk) [reply docker-compose.prod.yml:30]: Created

File: docker-compose.prod.yml
Comment (jamilbk) [inline docker-compose.prod.yml:35]: Maybe telemetry ID could be persisted here?

File: Dockerfile.prod
Comment (jamilbk) [inline Dockerfile.prod]: Should probably remain a private ENV var -- not overridable.

File: docker-compose.prod.yml
Comment (jamilbk) [inline docker-compose.prod.yml]: Should probably remain a private ENV var

**Review by jamilbk — APPROVED** [review]
This looks good, but we're gonna need docs. Another PR or this one?

**Comment by princemaple** [comment]
Was hoping that you could also approve my PR description 😉 if that looks okay I will move it into a markdown file.

**Pull request merged**
